File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* prev = dummy;
        ListNode* curr = dummy->next;
        while (true) {
            ListNode* kNode = prev;
            for (int i = 0; i < k; ++i) {
                kNode = kNode->next;
                if (!kNode) {
                    return dummy->next;
                }
            }
            ListNode* prevGroupEnd = prev;
            ListNode* nextGroupStart = kNode->next;
            while (curr != nextGroupStart) {
                ListNode* next = curr->next;
                curr->next = prev;
                prev = curr;
                curr = next;
            }
            ListNode* currGroupStart = prevGroupEnd->next;
            prevGroupEnd->next = prev;
            currGroupStart->next = nextGroupStart;
            prev = currGroupStart;
        }
        return dummy->next;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-4.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {void}
     */
    reorderList(head) {
        // Get center node
        let slow = head, fast = head.next;
        while (fast && fast.next) {
            slow = slow.next;
            fast = fast.next.next;
        }
        // Reverse list
        let node = slow.next;
        let prev = slow.next = null;
        while (node) {
            const node_next = node.next;
            node.next = prev;
            prev = node;
            node = node_next;
        }
        // Cross linked for reordering list
        let first = head, second = prev;
        while (second) {
            const node = second.next;
            second.next = first.next;
            first.next = second;
            first = second.next;
            second = node;
        }
    }
}
